### DISCUSSION TITLE
Adding a field in the MOOSE output exodus file using Python

### DISCUSSION THREAD

**Discussion by chakra34** [idaholab/moose#26752] [Q&A General] [answered]
Hi,

I was wondering if anyone have some suggestion in using Python to read the exodus output of MOOSE and then add extra fields for visualization in ParaView. For example I have stress values that are output for each grid point in ParaView and I want to scale them with different values based on some temperature dependent function. 

So my objective is to 

read the moouse_out.exo
get stress components
get temperature
add field = sigma_xx/(A*temperature +B)
write into exodus output

Any help would be great!! 

thanks,
Aritra

**Comment by GiudGiud** [comment]
Hello

You can work in Python with the VTK libraries to do all this. MOOSE can output to vtk

But it would be much easier to do this in a small input in MOOSE that just does not solve. Add this line to avoid solving
```
[Problem]
  solve = false
[]
```
and add an auxkernel (maybe a ParsedAux?) to compute the new field. You may need an auxkernel to get sigma_xx into an auxiliary variable first

> get stress components

that can come from auxkernels or from automated material property output (using the Material or the TensorMechanics Master parameters) in your simulation

Guillaume

**Reply by chakra34** [reply]
Thanks a lot :)

**Reply by chakra34** [reply]
@GiudGiud a quick follow up

So in my case, I simulate a linear elastic problem under a temperature field. I have the stresses in aux variable, and I supply the scaling function too (as xy-data; x-being temperature; y-being scaling factor)

Now, I want to add an auxvariable that would do the following: stress_xx/stress_scaling_func(temp) (essentially scaling the stress at the location with the scale factor that is a function of the temperature at that location.

Is the above possible via auxkernels and auxvariables?

I am attaching a simple moose_input file where I have put the "stress_scaling_func"

```
[Mesh]
  type = GeneratedMesh
  dim = 3
  nx = 20
  ny = 20
  nz = 20
  xmin = 0.0
  xmax = 1.0
  ymin = 0.0
  ymax = 1.0
  zmin = 0.0
  zmax = 1.0
[]

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]


[Functions]
  [./topfunc]
    type = ParsedFunction
    value = '1e-4*t' # ramped displacement BC
  [../]

  [./temp_function]
    type  = ParsedFunction
    value = '200.0*(z+0)'
  [../]

  **'comment:  x data is temperature ; y data is scaling_factor'**

  [./stress_scaling_func]
    type = PiecewiseLinear
    xy_data = '25	130.70
               500 1000.0'
    scale_factor = 1.0
  [../]
     
[]

[BCs]

  [./bot_fix_y]
    type = DirichletBC
    variable = disp_y
    boundary = 'bottom'
    value = 0
  [../]

  [./top_disp_y]
    type = FunctionDirichletBC
    variable = disp_y
    boundary = 'top'
    function = topfunc
  [../]

  [./left_fix_x]
    type = DirichletBC
    variable = disp_x
    boundary = 'left'
    value = 0
  [../]

  [./front_fix_z]
    type = DirichletBC
    variable = disp_z
    boundary = 'front'
    value = 0
  [../]

[]

[Modules/TensorMechanics/Master/All]
  strain = FINITE
  incremental = true
  add_variables = true
  temperature = temp
  eigenstrain_names = eigenstrain  
  generate_output = 'stress_xy stress_yz stress_zx'
  use_automatic_differentiation = false
[]
[Materials]
  [elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 150e9
    poissons_ratio = 0.3
  []

  [./stress]
    type = ComputeFiniteStrainElasticStress
  [../]

  [./thermal_expansion_strain]
    type = ComputeThermalExpansionEigenstrain
    thermal_expansion_coeff = 1.0e-4
    stress_free_temperature = 100.0
    temperature = temp
    eigenstrain_name = eigenstrain
  [../]

[]


[AuxVariables]
  [temp]
    order  = FIRST
    family = LAGRANGE
  []
  [./von_mises_stress]
	#Dependent variable used to visualize the Von Mises stress
    order = SECOND
    family = MONOMIAL
  [../]

  [./stress_yy]
    order = SECOND
    family = MONOMIAL
  [../]
  [./stress_xx]
    order = SECOND
    family = MONOMIAL
  [../]

[]

[AuxKernels]
  [set_temp]
     variable = temp
     type = FunctionAux
     function = temp_function
  []

  [./stress_yy]
    type = RankTwoAux
    rank_two_tensor = stress
    variable = stress_yy
    index_i = 1
    index_j = 1
    execute_on = 'initial timestep_end'
  [../]
  [./stress_xx]
    type = RankTwoAux
    rank_two_tensor = stress
    variable = stress_xx
    index_i = 0
    index_j = 0
    execute_on = 'initial timestep_end'
  [../]
  
[]



[Preconditioning]
  [smp]
    type = SMP
    full = true
  []
[]

[Executioner]
  type = Transient

  solve_type = 'Newton'

  start_time = 0.0
  dt    = 0.1
  dtmax = 1.0
  dtmin = 1e-31
  nl_abs_tol = 1e-6
  end_time = 1
 []

[Outputs]
  print_linear_residuals = true
  perf_graph = true
  csv = true
  [./out]
    type = Exodus
    elemental_as_nodal = true
  [../]
  [./console]
    type = Console
    max_rows = 100
  [../]
[]

```

Thanks,
Aritra

**Reply by GiudGiud** [reply]
hello

you mention an auxiliary variable for this. Is this for output? Or will it be used in the simulation?

Guillaume

**Reply by chakra34** [reply]
this is only output [ basically an extra output field of scaled value of stress_xx by the function which is a function of temperature at the element.]


Thanks again,
Aritra

**Reply by GiudGiud** [reply]
So yes an auxvariable would do great.
You already have stress_xx in an auxvariable
and stress_scaling_func in a function

so now use a ParsedAux to do the division of one by the other

**Reply by chakra34** [reply]
this is giving error

"""
  [scaling_stress]
    type = ParsedAux
    variable = scaled_stress_xx
    coupled_variables = 'stress_xx stress_scaling_func '
    function = 'stress_xx/stress_scaling_func'
  []
  
"""
*** ERROR ***
The following error occurred in the object "scaling_stress", of type "ParsedAux".

Invalid function
stress_xx
in ParsedAux scaling_stress.
Syntax error: Unknown identifier

"""

I tried the following and it still gave me Unknown identifier error:

"""
  [shear_stress_var]
    order = CONSTANT
    family = MONOMIAL
  []
"""
"""
  [shear_stress]
    type = ParsedAux
    variable = shear_stress_var
    coupled_variables = 'stress_yz stress_xz'
    function = 'sqrt(stress_yz^2 + stress_zx^2)'
  []
"""

the example in ParsedAux documentation.

Is there I am missing ?

Aritra

**Reply by GiudGiud** [reply]
is stress_xx in the input file still as an auxiliary variable?

**Reply by chakra34** [reply]
@GiudGiud I figured out the error, I am using an older version where instead of "coupled_variables" the tag is "args". Apologies for that.
But I have the following question-- so I have the scaling function:

```
  [./stress_scaling_func]
    type = PiecewiseLinear
    xy_data = '25	130.70
               500 1000.0'
    scale_factor = 1.0
  [../]
```
where the x is temperature and y is data. I made an auxvariable out of it by:

```
  [set_mor_a]
    variable = temp_mor_a
    type = FunctionAux
    function = stress_scaling_func
 []
```
And then I do the following parsedAux

```
  [shear_stress]
      type = ParsedAux
      variable = shear_stress_var
      args = 'stress_xx temp_mor_a'
      function = 'stress_xx/temp_mor_a'
    []
 ```

The code runs, but my question is how/where does the code know to use the temperature to get the temp_mor_a value which corresponds to the element location where the stress_xx is also evaluated ?

[As in in the above function the temp_mor_a would be ```stress_xx(x,y,z)/temp_mor_a(temperature(x,y,z))``` where (x,y,z) are the local coordinates/location.]

Best,
Aritra

**Reply by GiudGiud** [reply]
>  the temperature to get the temp_mor_a value which corresponds to the element location where the stress_xx is also evaluated ?

So the expression will be evaluated on quadrature points for. both the variable and the function, then using this a projection is made onto the target auxiliary variable.

**Reply by chakra34** [reply]
okay, thanks, sorry a quick follow up:  so while evaluating ```temp_mor_a``` on quadrature point how will it decide what to use as the input for the function. 

So let's say at the quadrature point my temperature is 500 C and displacement is 0.5, so how do we make sure when we evaluate the ```temp_mor_a``` which is a function using the temperature it uses the temperature values (500 C) and not other variables (e.g. displacement and 0.5) ?

In my case temperature is also an AuxVariable coming from the function:

```
  [./temp_function]
    type  = ParsedFunction
    value = '200.0*(z+0)'
  [../]

```
and the AuxVariable and AuxKernel being:

```
  [temp]
    order  = FIRST
    family = LAGRANGE
  []
```

```
  [set_temp]
     variable = temp
     type = FunctionAux
     function = temp_function
  []
```

I am really sorry if these questions are too trivial :( , the reason I am wondering is because in no place for the ```stress_scaling_func``` or the ParsedAux did we input the information of the temperature (AuxVariable) ```temp```.

Again, really appreciate the help @GiudGiud 

Aritra

**Reply by GiudGiud** [reply]
> how will it decide what to use as the input for the function.

it will use the local quadrature point coordinates and the current simulation time.
If the mesh is displaced, it will use the displaced mesh if the consumer of the function value is acting on the displaced mesh (iirc)

The temperature function here is being used to fill a temp variable field for the first lagrange family. We can get an exact representation by simply evaluating the function at the nodes here (and setting that as the DOF value for the variable), and that s what MOOSE does iirc.

**Reply by chakra34** [reply]
Thanks @GiudGiud . So I ran a case with the above, but the ```temp_mor_a``` auxvariable seems to be constant in the geometry:

This is my full MOOSE input file
```
[Mesh]
  type = GeneratedMesh
  dim = 3
  nx = 20
  ny = 20
  nz = 20
  xmin = 0.0
  xmax = 1.0
  ymin = 0.0
  ymax = 1.0
  zmin = 0.0
  zmax = 1.0
[]

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]


[Functions]
  [./topfunc]
    type = ParsedFunction
    value = '1e-4*t' # ramped displacement BC
  [../]

  [./temp_function]
    type  = ParsedFunction
    value = '200.0*(z+0)'
  [../]

  # x data is temperature ; y data is scaling_factor
  [./stress_scaling_func]
    type = PiecewiseLinear
    xy_data = '0.0	100.0
               200  200.0'
    scale_factor = 1.0
  [../]
     
[]

[BCs]

  [./bot_fix_y]
    type = DirichletBC
    variable = disp_y
    boundary = 'bottom'
    value = 0
  [../]

  [./top_disp_y]
    type = FunctionDirichletBC
    variable = disp_y
    boundary = 'top'
    function = topfunc
  [../]

  [./left_fix_x]
    type = DirichletBC
    variable = disp_x
    boundary = 'left'
    value = 0
  [../]

  [./front_fix_z]
    type = DirichletBC
    variable = disp_z
    boundary = 'front'
    value = 0
  [../]

[]

[Modules/TensorMechanics/Master/All]
  strain = FINITE
  incremental = true
  add_variables = true
  temperature = temp
  eigenstrain_names = eigenstrain  
  generate_output = 'stress_xy'
  use_automatic_differentiation = false
[]
[Materials]
  [elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 150e9
    poissons_ratio = 0.3
  []

  [./stress]
    type = ComputeFiniteStrainElasticStress
  [../]

  [./thermal_expansion_strain]
    type = ComputeThermalExpansionEigenstrain
    thermal_expansion_coeff = 1.0e-4
    stress_free_temperature = 100.0
    temperature = temp
    eigenstrain_name = eigenstrain
  [../]

[]


[AuxVariables]
  [temp]
    order  = FIRST
    family = LAGRANGE
  []
  [./von_mises_stress]
	#Dependent variable used to visualize the Von Mises stress
    order = SECOND
    family = MONOMIAL
  [../]

  [./stress_yy]
    order = SECOND
    family = MONOMIAL
  [../]
  [./stress_xx]
    order = SECOND
    family = MONOMIAL
  [../]

  [temp_mor_a]
    order  = FIRST
    family = LAGRANGE
  []
  
  [shear_stress_var]
    order = CONSTANT
    family = MONOMIAL
  []
      
[]

[AuxKernels]
  [set_temp]
     variable = temp
     type = FunctionAux
     function = temp_function
  []

  [set_mor_a]
    variable = temp_mor_a
    type = FunctionAux
    function = stress_scaling_func
 []

  [./stress_yy]
    type = RankTwoAux
    rank_two_tensor = stress
    variable = stress_yy
    index_i = 1
    index_j = 1
    execute_on = 'initial timestep_end'
  [../]
  [./stress_xx]
    type = RankTwoAux
    rank_two_tensor = stress
    variable = stress_xx
    index_i = 0
    index_j = 0
    execute_on = 'initial timestep_end'
  [../]

    [shear_stress]
      type = ParsedAux
      variable = shear_stress_var
      args = 'stress_xx temp_mor_a'
      function = 'stress_xx/temp_mor_a'
    []
  

[]



[Preconditioning]
  [smp]
    type = SMP
    full = true
  []
[]

[Executioner]
  type = Transient

  solve_type = 'Newton'

  start_time = 0.0
  dt    = 0.1
  dtmax = 1.0
  dtmin = 1e-31
  nl_abs_tol = 1e-6
  end_time = 0.2
 []

[Outputs]
  print_linear_residuals = true
  perf_graph = true
  csv = true
  [./out]
    type = Exodus
    elemental_as_nodal = true
  [../]
  [./console]
    type = Console
    max_rows = 100
  [../]
[]

```

and the following is the solution:

![image](https://github.com/idaholab/moose/assets/22983992/0fc59801-8034-4b4a-8f6a-e60b5fb14b4c)


As you see I have a temperature field but the ```temp_mor_a``` is constant.

Best,
Aritra

**Reply by GiudGiud** [reply]
Is this plotted at t=0? This FunctionAux is not executed on initial
Also in the function
```
  # x data is temperature ; y data is scaling_factor
  [./stress_scaling_func]
    type = PiecewiseLinear
    xy_data = '0.0	100.0
               200  200.0'
    scale_factor = 1.0
  [../]
```
you need to make sure the XY_data input is correct. 
What spatial dependence are you trying to have ? at x=0: f = 200 and at x=100, f = 200?

**Reply by chakra34** [reply]
Thanks @GiudGiud  for the reply.

No this is plotted at the end of the simulation. 
I want to have it dependent on the temperature, instead of space. So basically, what I want is the function to have:

x_data = temperature = 0.0 200.0 
y_data = value =          100.0 200.0

Since I have a temperature aux variable ```temp``` I want to use that to evaluate the value of this function and store it another aux variable ```temp_mor_a```.

So in this case the ```temp_mor_a``` should be similar to the ```temp``` field in that at  ```z=0.0 ; temp = 0.0 ; temp_mor_a = 100.0``` and at ```z=1.0; temp = 200.0 ; temp_mor_a = 200.0``` but as you see ```temp_mor_a``` is not using the ```temp``` to evaluate it's function value.

Best,
Aritra

**Reply by GiudGiud** [reply]
Functions do not depend on variables. They depend on space and time. 
You want to create either a material property for the scaling factor (or the output quantity) or an auxiliary variable.

**Reply by GiudGiud** [reply]
Since your function is very simple, a ParsedAux should be able to have the dependency you need

**Reply by chakra34** [reply]
Thanks @GiudGiud. The example I share is indeed quite simple, but the real case I am working on is basically an ASCII data of Temperature and values, which cannot be converted into a functional form in space. And my objective was to scale the stress_xx (for example) at the end of a MOOSE simulation with these values.

That's why I initially had thought of doing such as post-processing via the MOOSE output.

Thanks again,

Best,
Aritra

**Reply by GiudGiud** [reply]
@dschwen do we have an auxkernel to apply a function to a variable? Should we add that?

I dont think we do so the solution here is going to be to make a new auxiliary kernel. See the tutorial for making a new object. 
in that new auxkernel, in the computeValue routine, you can apply feed to the function the local value of the variable

**Reply by chakra34** [reply]
Got it! I will check out adding an AuxKernel! Thanks !

**Reply by chakra34** [reply]
Hi @GiudGiud , is there an existing AuxKernel that I can look into? FunctionAux ?

**Reply by GiudGiud** [reply]
FunctionAux sets an auxiliary variable to f(x,y,z,t) (the function) everywhere locally. So not quite what you want

You do need a make a new kernel

**Reply by chakra34** [reply]
hmm okay, thanks @GiudGiud . what I was thinking is this new auxKernel 

-- should get a function that is defined in the input (either in algebraic form or as ASCII table xy data)  (similar to FunctionAux)
-- But while getting the function value it will take input an auxvariable (that will one of the required argument) which is defined in each _qp


So, the input to this new Auxkernel would be
-- Function
-- auxvariable/field which will be used to evaluate the function at _qp

And output 
-- Auxvariable of this new field 

So in that sense shouldn't it be similar to FunctionAux.C partly in the following

```
InputParameters
FunctionAux::validParams()
{
  InputParameters params = AuxKernel::validParams();
  params.addClassDescription("Auxiliary Kernel that creates and updates a field variable by "
                             "sampling a function through space and time.");
  params.addRequiredParam<FunctionName>("function", "The function to use as the value");
  return params;
}

FunctionAux::FunctionAux(const InputParameters & parameters)
  : AuxKernel(parameters), _func(getFunction("function"))
{
}

```
I am thinking, ```_func``` should just create the function based on the input we provide right ? Does it inherently assume space and time dependency in some way ?

If that is the case then 

I can modify the following 

```
Real
FunctionAux::computeValue()
{
  if (isNodal())
    return _func.value(_t, *_current_node);
  else
    return _func.value(_t, _q_point[_qp]);
}

```

to something like

```
Real
MyNewFunctionAux::computeValue()
{
  if (isNodal())
    return _func.value(_input_aux_variable[_qp], *_current_node);
  else
    return _func.value(_input_aux_variable[_qp], _q_point[_qp]);
}

```

Does the above make sense? or am I completely off with the logic


Best,
Aritra

**Reply by GiudGiud** [reply, marked as answer]
yup this looks good. 
So here you are using the 'input-aux-variable' as the 'time' variable in the function

**Reply by chakra34** [reply]
Hi @GiudGiud , I wrote an AuxKernel and it works as mentioned above. so closing the question. Thanks for all the help! :) 

Aritra
